xocto/ranges.py

```diff
@@ -1034,3 +1038,34 @@ def as_finite_datetime_periods(
         finite_periods += [FiniteDatetimeRange(period.start, period.end)]
 
     return finite_periods
+
+
+def iterate_over_months(
+    period: FiniteDateRange, tz: datetime.tzinfo
+) -> Iterator[FiniteDatetimeRange]:
+    """
+    Generate a sequence of finite datetime ranges spanning months during the period.
+
+    Ranges span each whole month between the start and end times, inclusive.
+
+    ie: given a period of (15/01, 15/03) -> [
+        (15/01->01/02),
+        (01/02->01/03),
+        (01/03->15/03)
+    ]
+
+    Datetimes will be aligned to midnight as at the supplied timezone.
+    """
+    start_at = localtime.midnight(period.start, tz=tz)
+    end_at = localtime.next_midnight(period.end, tz=tz)
```

Comment:
What happens if `end_at` is already midnight?

Reply:
I had made the `period` argument a `DateRange` instead of a `DatetimeRange` since we're rounding to midnight anyway, but thinking about it further that would then probably force people to convert datetimes to dates when calling this, which is annoying.

I've now changed `period` to be a `DatetimeRange`, and changed this line to `end_at = localtime.next_midnight(period.end, tz=tz)`

Reply:
I think this is safe now.